sklearn/ensemble/_forest.py

```diff
@@ -870,11 +854,34 @@ def _validate_y_class_weight(self, y):
                         "class_weight parameter."
                     )
 
-            if self.class_weight != "balanced_subsample" or not self.bootstrap:
-                if self.class_weight == "balanced_subsample":
-                    class_weight = "balanced"
-                else:
-                    class_weight = self.class_weight
+            # User defined class_weight (dict or list)
+            if isinstance(self.class_weight, (dict, list)):
+                class_weight = self.class_weight
+            # Computing class_weight (dict or list) for the "balanced" option.
+            # The "balanced_subsample" option without subsampling (bootstrap=False)
+            # is equivalent to the "balanced" option.
+            elif (self.class_weight == "balanced") or (
+                self.class_weight == "balanced_subsample" and not self.bootstrap
+            ):
+                class_weight = []
+                for k in range(self.n_outputs_):
+                    class_weight_k_vect = compute_class_weight(
+                        "balanced",
+                        classes=self.classes_[k],
+                        y=y_original[:, k],
+                        sample_weight=sample_weight,
+                    )
+                    class_weight_k = {
+                        key: val
+                        for (key, val) in zip(self.classes_[k], class_weight_k_vect)
+                    }
+                    class_weight.append(class_weight_k)
+                if self.n_outputs_ == 1:
+                    class_weight = class_weight[0]
+
+            # For the "balanced_subsample" option with subsampling (bootstrap=True),
+            # class_weight will be computed on the bootstrap sample for each tree.
```

Comment:
I almost wonder if it is worth doing an early return up in line 817 where we do `expanded_class_weight = None`, and then this comment would be easier to understand (as it is easy to forget we set `expanded_class_weight` to None above)?